pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -625,24 +648,47 @@ func (vs *VSphere) AttachDisk(vmDiskPath string, storagePolicyID string, nodeNam
 		if err != nil {
 			return "", err
 		}
-		vm, err := vs.getVMByName(ctx, nodeName)
+
+		vm, err := vs.getVMFromNodeName(ctx, nodeName)
 		if err != nil {
 			glog.Errorf("Failed to get VM object for node: %q. err: +%v", convertToString(nodeName), err)
 			return "", err
 		}
-		diskUUID, err = vm.AttachDisk(ctx, vmDiskPath, &vclib.VolumeOptions{SCSIControllerType: vclib.PVSCSIControllerType, StoragePolicyID: storagePolicyID})
+
+		diskUUID, err = vm.AttachDisk(ctx, vmDiskPath, &vclib.VolumeOptions{SCSIControllerType: vclib.PVSCSIControllerType, StoragePolicyName: storagePolicyName})
 		if err != nil {
 			glog.Errorf("Failed to attach disk: %s for node: %s. err: +%v", vmDiskPath, convertToString(nodeName), err)
 			return "", err
 		}
 		return diskUUID, nil
 	}
 	requestTime := time.Now()
-	diskUUID, err = attachDiskInternal(vmDiskPath, storagePolicyID, nodeName)
+	diskUUID, err = attachDiskInternal(vmDiskPath, storagePolicyName, nodeName)
+	if err != nil {
+		isManagedObjectNotFoundError, err := vs.retry(nodeName, err)
+		if isManagedObjectNotFoundError {
+			if err == nil {
+				glog.V(4).Infof("AttachDisk: Found node %q", convertToString(nodeName))
+				diskUUID, err = attachDiskInternal(vmDiskPath, storagePolicyName, nodeName)
+			}
+		}
+	}
 	vclib.RecordvSphereMetric(vclib.OperationAttachVolume, requestTime, err)
 	return diskUUID, err
 }
 
+func (vs *VSphere) retry(nodeName k8stypes.NodeName, err error) (bool, error) {
```

Comment:
Can we rename this method? Maybe call it as checkNodeNotPresentAndRediscover or isManagedObjectNotFound or checkForRediscovery or checkVmNotFound.